store/sqlstore/thread_store.go

```diff
@@ -419,44 +419,68 @@ func (s *SqlThreadStore) GetThreadForUser(teamId string, threadMembership *model
 	return result, nil
 }
 
-// MarkAllAsReadInChannels marks all threads for the given user in the given channels as read from
-// the current time.
-func (s *SqlThreadStore) MarkAllAsReadInChannels(userID string, channelIDs []string) error {
-	threadIDs := []string{}
-
-	query, args, _ := s.getQueryBuilder().
-		Select("ThreadMemberships.PostId").
-		Join("Threads ON Threads.PostId = ThreadMemberships.PostId").
-		Join("Channels ON Threads.ChannelId = Channels.Id").
-		From("ThreadMemberships").
+// MarkAllAsReadByChannels marks thread membership for the given users in the given channels
+// as read. This is used by the application layer to keep threads up-to-date when CRT is disabled
+// for the enduser, avoiding an influx of unread threads when first turning the feature on.
+//
+// Note that an alternate implementation of this method using UPDATE FROM actually performed worse
+// than the subquery with the extra JOIN below. Since the query constrains itself only to threads
+// which have been updated since the last time the channel was viewed, the subquery outperforms
+// the UPDATE FROM as it forces a lookup of a small number of post ids (on average) instead of
+// using other indexes less efficiently to reach the same conclusion.
+func (s *SqlThreadStore) MarkAllAsReadByChannels(userID string, channelIDs []string) error {
+	now := model.GetMillis()
+
+	subquery, subqueryArgs, _ := s.getQueryBuilder().
+		PlaceholderFormat(sq.Question).
+		Select("Threads.PostId").
+		From("Threads").
+		Join("ThreadMemberships ON ThreadMemberships.PostId = Threads.PostId").
 		Where(sq.Eq{"Threads.ChannelId": channelIDs}).
 		Where(sq.Eq{"ThreadMemberships.UserId": userID}).
+		Where(sq.Expr("Threads.LastReplyAt > ThreadMemberships.LastViewed")).
 		ToSql()
 
-	err := s.GetReplicaX().Select(&threadIDs, query, args...)
-	if err != nil {
-		return errors.Wrapf(err, "failed to get thread membership with userid=%s", userID)
+	if s.DriverName() == model.DatabaseDriverMysql {
+		// Workaround #1093 - You can't specify target table 'xxx' for update in FROM clause
+		subquery = "SELECT PostId FROM (" + subquery + ") as t"
 	}
 
-	timestamp := model.GetMillis()
-	query, args, _ = s.getQueryBuilder().
+	query, args, _ := s.getQueryBuilder().
 		Update("ThreadMemberships").
-		Where(sq.Eq{"PostId": threadIDs}).
 		Where(sq.Eq{"UserId": userID}).
+		Where(sq.Expr("PostId IN ("+subquery+")", subqueryArgs...)).
+		Set("LastViewed", now).
```

Comment:
Should we set this to `Threads.LastReplyAt` ? `Now` could be a lot after `Threads.LastReplyAt`. If there are network/lag issues (mobile client), could we conceivably be marking a thread as read for posts that are still in-flight? 

This is probably not an issue for a user with CRT turned off. Users with two clients, one of which has CRT and the other is unsupported, could see these still in flight posts as read after switching from the CRT off to the CRT on client. 

Analogous to why the the channelMembership.LastViewedAt isn't set to the anything greater than the CreateAt of the latest post. 

See - channel_store:UpdateLastViewedAt and https://community-daily.mattermost.com/core/pl/ff87f31mfb85pfipnw1sqgx81e

Reply:
Hmm, this would be an option, but I expect it to cost us significantly more -- likely need a CTE a pair of queries like we do for the MySQL case.

How often do we expect this to happen? Are there still official clients that don't support CRT, or was this mostly historical? Are there third-party clients (like Matterhorn) that we're considering here?

I'd be inclined to keep this query as simple as possible if we want to preserve the observed performance gains, and wonder if we might be willing to sacrifice the risk of an in-flight message in such cases.

Reply:
The oldest supported server version is the current ESR 5.37. The oldest compatible mobile client is v1.47. CRT was added to mobile in v1.46.
The CRT user setting is a user level setting, not a client level setting. This means the user has CRT on on all clients or no clients. Since this code will only run with CRT off, it's not going to be a problem when switching clients. 
Third party clients that don't support CRT could be a problem. 

Given all of the above, I'm ok with keeping this query simple for performance reasons.